pkg/reader/usdc_reader.go

```diff
@@ -0,0 +1,308 @@
+	// We need entire MessageSent payload to use that with the Attestation API
+	cr, ok := u.boundContracts[source]
+	if !ok {
+		return nil, fmt.Errorf("no contract bound for chain %d", source)
+	}
+
+	iter, err := u.contractReaders[source].QueryKey(
+		ctx,
+		cr,
+		query.KeyFilter{
+			Key:         consts.EventNameCCTPMessageSent,
+			Expressions: []query.Expression{},
+		},
+		query.NewLimitAndSort(query.Limit{}, query.NewSortBySequence(query.Asc)),
+		&MessageSentEvent{},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error querying contract reader for chain %d: %w", source, err)
+	}
+
+	messageSentEvents := make(map[eventID]MessageHash)
+	for _, item := range iter {
+		event, ok1 := item.Data.(*MessageSentEvent)
+		if !ok1 {
+			return nil, fmt.Errorf("failed to cast %v to Message", item.Data)
+		}
+		e, err1 := event.unpackID()
+		if err1 != nil {
+			return nil, err1
+		}
+		messageSentEvents[e] = event.Arg0
+	}
+
+	// 3. This should be done by ChainReader - picking only events matching eventIDs
+	out := make(map[exectypes.MessageTokenID]MessageHash)
+	for tokenID, messageID := range eventIDs {
+		messageHash, ok1 := messageSentEvents[messageID]
+		if !ok1 {
+			// Token not available in the source chain
+			continue
+		}
+		out[tokenID] = messageHash
+	}
+
+	return out, nil
+}
+
+func (u usdcMessageReader) recreateMessageTransmitterEvents(
+	destChainSelector cciptypes.ChainSelector,
+	tokens map[exectypes.MessageTokenID]cciptypes.RampTokenAmount,
+) (map[exectypes.MessageTokenID]eventID, error) {
+	messageTransmitterEvents := make(map[exectypes.MessageTokenID]eventID)
+
+	for id, token := range tokens {
+		sourceTokenPayload, err := NewSourceTokenDataPayloadFromBytes(token.ExtraData)
+		if err != nil {
+			return nil, err
+		}
+
+		// USDC message payload:
+		// uint32 _msgVersion,
+		// uint32 _msgSourceDomain,
+		// uint32 _msgDestinationDomain,
+		// uint64 _msgNonce,
+		// bytes32 _msgSender,
+		// Since it's packed, all of these values contribute to the first slot
+
+		msgVersionBytes := [4]byte{}
+		binary.BigEndian.PutUint32(msgVersionBytes[:], CCTPMessageVersion)
+
+		sourceDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(sourceDomainBytes[:], sourceTokenPayload.SourceDomain)
+
+		destDomain, ok := u.cctpDestDomain[uint64(destChainSelector)]
+		if !ok {
+			return nil, fmt.Errorf("destination domain not found for chain %s", destChainSelector)
+		}
+
+		destDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(destDomainBytes[:], destDomain)
+
+		nonceBytes := [8]byte{}
+		binary.BigEndian.PutUint64(nonceBytes[:], sourceTokenPayload.Nonce)
+
+		senderBytes := [12]byte{}
+
+		messageTransmittedEvent := [32]byte(
+			append(
+				append(
+					append(
+						append(
+							msgVersionBytes[:],
+							sourceDomainBytes[:]...),
+						destDomainBytes[:]...),
+					nonceBytes[:]...),
+				senderBytes[:]...),
+		)
+		messageTransmitterEvents[id] = messageTransmittedEvent
```

Comment:
The nested appends were a bit confusing for me, suggest byte buffer: 
```
		var buf bytes.Buffer
		buf.Write(msgVersionBytes[:])
		buf.Write(sourceDomainBytes[:])
		buf.Write(destDomainBytes[:])
		buf.Write(nonceBytes[:])
		buf.Write(senderBytes[:])
		messageTransmitterEvents[id] = buf.Bytes()[0:32]
```

Or if you want to be precise about the packing:
```
		messageTransmittedEvent := [32]byte{}
		copy(messageTransmittedEvent2[:], msgVersionBytes[:])
		copy(messageTransmittedEvent2[4:], sourceDomainBytes[:])
		copy(messageTransmittedEvent2[8:], destDomainBytes[:])
		copy(messageTransmittedEvent2[12:], nonceBytes[:])
		copy(messageTransmittedEvent2[20:], senderBytes[:])
		messageTransmitterEvents[id] = messageTransmittedEvent
```
https://play.golang.com/p/F3t5uBny_VJ